src/transformers/training_args.py

```diff
@@ -807,6 +813,13 @@ def __post_init__(self):
                 )
             if not (self.sharded_ddp == "" or not self.sharded_ddp):
                 raise ValueError("sharded_ddp is not supported with bf16")
+
+        if self.adafactor:
+            warnings.warn(
+                "`adafactor` is deprecated and will be removed in version 5 of 🤗 Transformers. Use `--optim adafactor` instead",
+                FutureWarning,
+            )
+
```

Comment:
right, so here, before the deprecation warning, you set `self.optim` to `adafactor` and simplify the logic in the trainer above in the PR

Reply:
Good point!

Reply:
Yes! This is much cleaner.